bugsnag-android-core/src/main/java/com/bugsnag/android/internal/MemoryMappedOutputStream.kt

```diff
@@ -0,0 +1,61 @@
+package com.bugsnag.android.internal
+
+import java.io.File
+import java.io.OutputStream
+import java.io.RandomAccessFile
+import java.nio.MappedByteBuffer
+import java.nio.channels.FileChannel
+
+/**
+ * OutputStream wrapper around a shared memory file.
+ *
+ * Memory-mapped files are maintained by the operating system, and so they are guaranteed to match
+ * the last data written to the memory region, even if the app crashes.
+ *
+ * The memory mapped file is fixed to the length specified at instantiation, meaning that
+ * write() will throw java.nio.BufferOverflowException if the file is too full.
+ */
+class MemoryMappedOutputStream(
+    file: File,
+    private val bufferSize: Int,
+    private val clearedByteValue: Byte
+) : OutputStream() {
+
+    private val raf: RandomAccessFile = RandomAccessFile(file, "rw")
+    private val memory: MappedByteBuffer =
+        raf.channel.map(FileChannel.MapMode.READ_WRITE, 0, bufferSize.toLong())
+
+    init {
+        raf.setLength(bufferSize.toLong())
+    }
+
+    fun clear() {
+        memory.rewind()
+        for (i in 1..bufferSize) {
+            memory.put(clearedByteValue)
+        }
+        memory.rewind()
+    }
+
+    override fun close() {
+        memory.force()
+        raf.close()
+    }
+
+    /**
+     * Write a byte value to the file.
+     * @throws java.nio.BufferOverflowException if the file is too full.
```

Comment:
What sort of strategy will the journal changes use to avoid the file getting too full? Presumably handling will take place higher up away from this class?

Reply:
They calling object will catch this exception and then do some housekeeping before retrying.

Reply:
```kotlin
    fun addCommand(command: Journal.Command) {
        if (!isOpen) throw IllegalStateException("Cannot add commands to a closed document")
        try {
            command.serialize(journalStream)
        } catch (ex: BufferOverflowException) {
            snapshot()
            command.serialize(journalStream)
        }
        command.apply(document)
        journal.add(command)
    }
```